statistics/histogram.go

```diff
@@ -826,12 +826,108 @@ func (hg *Histogram) AvgCountPerNotNullValue(totalCount int64) float64 {
+//           /x│  │◄─histogram─►│    \
+//         / xx│  │    range    │      \
+//       / │xxx│  │             │        \
+//     /   │xxx│  │             │          \
+//────┴────┴───┴──┴─────────────┴───────────┴─────
+//    ▲    ▲   ▲  ▲             ▲           ▲
+//    │    │   │  │             │           │
+// boundL  │   │histL         histR       boundR
+//         │   │
+//    lDatum  rDatum
+func (hg *Histogram) outOfRangeRowCount(lDatum, rDatum *types.Datum, increaseCount int64) float64 {
+	if hg.Len() == 0 {
+		return 0
+	}
+
+	// For bytes and string type, we need to cut the common prefix when converting them to scalar value.
+	// Here we calculate the length of common prefix.
+	commonPrefix := 0
+	if hg.GetLower(0).Kind() == types.KindBytes || hg.GetLower(0).Kind() == types.KindString {
+		// Calculate the common prefix length among the lower and upper bound of histogram and the range we want to estimate.
+		commonPrefix = commonPrefixLength(hg.GetLower(0).GetBytes(), hg.GetUpper(hg.Len()-1).GetBytes())
+		commonPrefix2 := commonPrefixLength(lDatum.GetBytes(), rDatum.GetBytes())
+		if commonPrefix2 < commonPrefix {
+			commonPrefix = commonPrefix2
+		}
+		commonPrefix3 := commonPrefixLength(hg.GetLower(0).GetBytes(), lDatum.GetBytes())
+		if commonPrefix3 < commonPrefix {
+			commonPrefix = commonPrefix3
+		}
+	}
+
+	// Convert the range we want to estimate to scalar value(float64)
+	l := convertDatumToScalar(lDatum, commonPrefix)
+	r := convertDatumToScalar(rDatum, commonPrefix)
+	// make sure l < r
+	if l >= r {
+		return 0
+	}
+	// Convert the lower and upper bound of the histogram to scalar value(float64)
+	histL := convertDatumToScalar(hg.GetLower(0), commonPrefix)
+	histR := convertDatumToScalar(hg.GetUpper(hg.Len()-1), commonPrefix)
+	histWidth := histR - histL
+	if histWidth <= 0 {
+		return 0
+	}
+	boundL := histL - histWidth
+	boundR := histR + histWidth
+
+	leftPercent := float64(0)
+	rightPercent := float64(0)
+
+	// keep l and r unchanged, use actualL and actualR to calculate.
+	actualL := l
+	actualR := r
+	// Handling the out-of-range part on the left of the histogram range
+	if actualL < histL && actualR > boundL {
+		// make sure boundL <= actualL < actualR <= histL
+		if actualL < boundL {
+			actualL = boundL
+		}
+		if actualR > histL {
+			actualR = histL
+		}
+		// Calculate the percentage of "the shaded area" on the left side.
+		leftPercent = (math.Pow(actualR-boundL, 2) - math.Pow(actualL-boundL, 2)) / math.Pow(histWidth, 2)
+	}
+
+	actualL = l
+	actualR = r
+	// Handling the out-of-range part on the right of the histogram range
+	if actualL < boundR && actualR > histR {
+		// make sure histR <= actualL < actualR <= boundR
+		if actualL < histR {
+			actualL = histR
+		}
+		if actualR > boundR {
+			actualR = boundR
+		}
+		// Calculate the percentage of "the shaded area" on the right side.
+		rightPercent = (math.Pow(boundR-actualL, 2) - math.Pow(boundR-actualR, 2)) / math.Pow(histWidth, 2)
```

Comment:
How about using `totalCountInHist/(histR-shitL)` as the `histHeigh` instead of assume `histHeight = histWidth`?